### ISSUE TITLE
jacobian not working when np.prod is used and one element is 0.

### ISSUE THREAD

**Issue by bigblindbais** [HIPS/autograd#93]
Hi, I am starting to use autograd for my own project and it's an amazing tool!

I encountered an issue very early during my tests, where the jacobian fails to be computed if I use the np.prod function, and only in the particular case when one entry is zero.  Here is a minimal working example of the problem:

```python
# import numpy as np
import autograd.numpy as np
from autograd import jacobian


def fun1(x):
    return np.array([x[0], x[1], x[0] * x[1]])


def fun2(x):
    return np.array([x[0], x[1], x.prod()])

fun1_jac = jacobian(fun1)
fun2_jac = jacobian(fun2)

x = np.array([0., 2.])  # when there is no zero element, both methods work.

print 'fun1:'
print fun1(x)
print fun1_jac(x)

print 'fun2:'
print fun2(x)
print fun2_jac(x)
```

I'm really new to this package, and only started using it yesterday, so I hope I have not misused or overlook any feature which would explain the above.

**Comment by mattjj** [comment]
Great find! I think this is a bug in the [gradient definition for `np.prod`](https://github.com/HIPS/autograd/blob/22285c1553c59eb1ed8647e674399765f10540f8/autograd/numpy/numpy_grads.py#L262-L265).

Specifically, I think [this line](https://github.com/HIPS/autograd/blob/22285c1553c59eb1ed8647e674399765f10540f8/autograd/numpy/numpy_grads.py#L264) assumes that `x != 0.` anywhere. That is, we're taking the gradient of something like `ans = x*y`, so just like in [`the gradient of __mul__`](https://github.com/HIPS/autograd/blob/22285c1553c59eb1ed8647e674399765f10540f8/autograd/core.py#L321) the gradient we return should be something like `g * y`. However, this code is effectively reconstructing `y` via `y = ans / x`, which is only valid where `x != 0`.

**Comment by mattjj** [comment]
I think the issue is basically like [this stackoverflow question](https://stackoverflow.com/questions/22435175/in-numpy-calculating-a-matrix-where-each-cell-contains-the-product-of-all-the-o), though I didn't write this code so I might not be thinking about it clearly yet.

**Comment by bigblindbais** [comment]
What you describe to be the problem seems to be spot on:  When one element is zero, then the method used to reconstruct partial derivative (i.e., smth like `np.prod(a) / a[i]`) results in 0 / 0, i.e. undefined.  The only way to correctly reconstruct the partial derivative in such case is to have access to all the elements in `a`.

**Comment by mattjj** [comment]
And we do have access to all of `x` in that code, but I can't think of a slick numpy way to construct the values we want (basically the same thing as discussed in that stackoverflow question). We want a numpy implementation not just for efficiency but also for closure under higher-order differentiation.

**Comment by dougalm** [comment]
Yep, this is definitely an autograd bug. Thanks for catching it. Our systematic tests missed it because they test random Gaussian inputs that never land on exactly zero. We should fix that.

We could use something like np.where to catch the places where x is exactly zero and substitute 1 instead (so 0/0 becomes 0/1). But it will still be numerically very close to zero.

**Comment by bigblindbais** [comment]
One simple way to achieve this would be to make a "copy" of the input array `x`, compute the repeated array (just as in the code), and then substitute all the "diagonal" elements with 1 before computing the result.  I'm figuring our a way to write this which is conformant to the rest of the code.

**Comment by mattjj** [comment]
@dougalm I tried `np.where` first actually (using `with np.errstate` and `np.isnan`), but that didn't work because the problem isn't just avoiding a `nan` when `0./0.` happens; it's that `y = ans / x` isn't valid, yet the `y` we want could be any nonzero value (which got thrown away when we multiplied by `x`, so `ans / x` doesn't recover it).

@bigblindbais Unfortunately, we can't use assignment. Also, I think any solution is going to incur a big overhead for when `x` has no nonzero values (though maybe we could just check for the fast case on the fly...).

**Comment by bigblindbais** [comment]
@mattjj: Yeah the code I posted was mostly to show how I thought we could do it in principle, it's not adjusted to the autograd framework.  If the overhead comment is a response to my numerical stability comment, what I meant is that we might want to avoid multiplying by a number and its inverse altogether, even with non-zero entries, not due to performance but to numerical stability. Granted, this in itself might be a non-issue altogether.

I know assignment won't work, but would the following?  It constructs a copy explicitly which is AFAIK possible in autograd:
```python
x = np.array([-1, 0, 1, 2])
y = np.array([[t for i, t in enumerate(x) if i != j] for j in range(x.size)])
print y.prod(axis=1)  # vector of partial derivatives of x.prod()
```

**Comment by mattjj** [comment]
Unfortunately, Python loops are slow, and we'd have to work to generalize that solution to arbitrary dimensions (remember, `x` could have any shape and the product could be along any axis).

Doing a bidirectional `np.cumprod` is probably the most asymptotically time- and memory-efficient way, and it's easy to pass in the `axis` argument there for generality (though I see we don't have its gradient implemented yet).
